### PR TITLE
Azure Driver update

### LINKED ISSUE #229: Manage azure resources with ARM template.
## Issue
Right now, we use multiple calls API to make atomic VM operations like create and delete

## Solution
Combine there create/delete NIC, Disk, and the VM in one ARM template.

### PR DESCRIPTION
**Pull request by chgeuer**
- Update Azure SDK (12.5-beta to 26.1)
- Create VM using idempotent ARM deployment. 
- Concurrent resource deletion (NIC/disk) where possible. 
- Ensure 'x-ms-request-id' header is logged in case of API failures.

Fixes #229

### CODE REVIEW

**Comment by hardikdr** [comment]
thanks @chgeuer for the PR. 
We will soon review it. :)

File: pkg/driver/driver_azure.go [hunk 1]
```diff
@@ -46,282 +48,416 @@ type AzureDriver struct {
...
-		nicName       = vmName + "-nic"
-		diskName      = vmName + "-os-disk"
-		location      = d.AzureMachineClass.Spec.Location
-		resourceGroup = d.AzureMachineClass.Spec.ResourceGroup
-		UserDataEnc   = base64.StdEncoding.EncodeToString([]byte(d.UserData))
+		ctx                  = context.Background()
+		vmName               = strings.ToLower(d.MachineName)
+		location             = d.AzureMachineClass.Spec.Location
+		resourceGroupName    = d.AzureMachineClass.Spec.ResourceGroup
+		deploymentName       = fmt.Sprintf("gardener-shoot-deployment-%s", vmName) // Deployments per resource group in the deployment history == 800
```
[the remark this answers is not in the document]
Comment (chgeuer) [reply pkg/driver/driver_azure.go]: You're right, that prefix can be removed.

**Comment by chgeuer** [comment]
You're right, the PR does not directly solve the deletion problems. By bringing it to the current Azure SDK and moving from imperative code to ARM deployment, we're now in good shape to move from availability set to VM Scale Set.

**Comment by vlerenc** [comment]
@chgeuer Where would you see the advantages of scale sets?

Let me explain our viewpoint: When we started the project, we used e.g. ASGs on AWS. They were working nicely and all that and certainly Azure scale sets will, too. But that’s not the point. We kicked that all out, because of the unneeded complexity and increased effort in trouble shooting. We actually don’t want to go back/what would be the benefit? Is direct VM management inferior (and why) to Azure scale sets? If not, we are quite reluctant.

- Scale sets are additional resources we need to manage, they certainly need some template how the VM looks like (in AWS speak we had ASGs and launch configurations). Today we don’t have these two additional resources anymore. So much cleaner without them.
- Consequently, since the scale out/in logic is anyways in cluster autoscaler and MCM, scale out/in got so much simpler. Add a VM/kill a particular VM, no awkward manipulation. So much cleaner without them.
- Adding the 1st machine or the Nth machine makes no difference in MCM today, same code for scale-out/in. With scale sets, you first need to manage those resources to get going when you create the cluster, but then the actual scale out/in operation is another piece code where you modify the desired count of the scale set (deletion even more complicated). So much cleaner without them.
- When ASGs were having trouble, we needed to check the ASG logs. They were hyperscaler-specific and not available in Gardener/MCM, because these are independent components. So much more hassle with them and less transparent for end users.
- Independent components bring unexpected behavior. So much more predictable without them.
- More hyperscaler-specific logic and ops effort with scale sets. So much cleaner without them.
 
Don’t get me wrong, when you manage scale sets manually or maybe only do one infrastructure, it’s certainly a good thing. But we need to have and already have a generic cluster autoscaler and machine management. There are really only disadvantages in going back to scale sets for us unless you can help us understand why.

**Comment by vlerenc** [comment]
I had a chat with @chgeuer on the phone. He recommends scale sets, because they are Microsofts/Azures recommendation for Kubernetes clusters on Azure (whatever the provider). If we do not use them, we basically do not follow the official recommendation. I am still not really sure, why it's not possible to achieve the same with basic resources like VM+Disk+NIC, but that's the official statement.

@chgeuer I would love to learn, why that's better. It would also nice to get your feedback on the list of issues we see with going back to scale sets (in general; this has nothing to do with Azure in particular).

In any case, if we really should go for scale sets, I don't think it makes sense to validate the intermediate solution with an ARM template for VM+Disk+NIC for creation and manual calls for deletion. Then it makes more sense to go directly/straight for an ARM template that creates (and can then also delete) the scale set.

Again, scale sets would complicate things as described in the bullet point list above, so feedback in regards to that list would be very welcome @chgeuer.

**Comment by dkistner** [comment]
I still see the issue that we are now introduce, by using ARM templates, a declarative way to define VMs and their related resources (NIC, System Disk), but the deletion of those resources in the same way is not supported yet, which means we need to delete the VMs + resources the old way.

If my understanding is correct then ARM templates could help at the moment to speed up the VM creation by relying on the Azure Resource Manager for the creation of those resources. ARM can internally parallelize the creation of the required resources. That's the main benefit/purpose of this pr, right?

@chgeuer I know your colleagues are working already on this selective deletion feature for ARM. Are you aware of a timeline in this regards?

I think scale sets are another discussion which we need to discuss separately.

**Comment by chgeuer** [comment]
Hi @vlerenc, regarding your questions on VM Scale Sets, this concrete PR is not related to integration of VM Scale Sets into Gardener. I understand your team had a technical discussion with Redmond colleagues the week before last week, so it might be good to sync with them. 

Hi @dkistner for the selective deletion I do not have a timeline, will check internally.

**Comment by prashanth26** [comment]
Hi @chgeuer,

I apologize for getting back on this topic after a long time. 

We had a few rounds of discussion internally and thought of sticking to traditional create calls for each resource instead of ARM templates for creation.

Having this in mind and avoid trouble for you, what I have done is taken your PR and rebased with current master and removed changes related to ARM templates. I hope that's alright. If it's not, you are free to make my changes on this PR and push your changes as well 👍 

New PR: https://github.com/gardener/machine-controller-manager/pull/261

It would be great if you were kind enough to review the new changes. TIA.

Kind Regards,
Prashanth

**Comment by prashanth26** [comment]
Closing in favor of https://github.com/gardener/machine-controller-manager/pull/261

**Pull request closed without merging**